sphinxgallery/gen_rst.py

```diff
@@ -292,157 +271,169 @@ def scale_image(in_fname, out_fname, max_width, max_height):
 
-    return image_list, time_elapsed, stdout
+def generate_dir_rst(src_dir, target_dir, gallery_conf, seen_backrefs):
+    """Generate the gallery reStructuredText for an example directory"""
+    if not os.path.exists(os.path.join(src_dir, 'README.txt')):
+        print(80 * '_')
+        print('Example directory %s does not have a README.txt file' %
+              src_dir)
+        print('Skipping this directory')
+        print(80 * '_')
+        return ""  # because string is an expected return type
+
+    fhindex = open(os.path.join(src_dir, 'README.txt')).read()
+    if not os.path.exists(target_dir):
+        os.makedirs(target_dir)
+    sorted_listdir = [fname for fname in sorted(os.listdir(src_dir))
+                                if fname.endswith('py')]
+    for fname in sorted_listdir:
+        generate_file_rst(fname, target_dir, src_dir)
+        new_fname = os.path.join(src_dir, fname)
+        intro = extract_intro(new_fname)
+        write_backreferences(seen_backrefs, gallery_conf,
+                             target_dir, fname, intro)
+
+        fhindex += _thumbnail_div(target_dir, fname, intro)
+        fhindex += """
+
+.. toctree::
+   :hidden:
+
+   /%s/%s\n""" % (target_dir, fname[:-3])
+
+
+# clear at the end of the section
+    fhindex += """.. raw:: html\n
+    <div style='clear:both'></div>\n\n"""
+
+    return fhindex
+
+
+def execute_script(code_block, example_globals, image_path, fig_count, src_file):
+    """Executes the code block of the example file"""
+    time_elapsed = 0
+    stdout = ''
+
+    # We need to execute the code
+    print('plotting %s' % src_file)
 
+    plt.close('all')
+    cwd = os.getcwd()
+    # Redirect output to stdout and
+    orig_stdout = sys.stdout
 
-def generate_file_rst(fname, target_dir, src_dir, plot_gallery):
+    try:
+        # First CD in the original example dir, so that any file
+        # created by the example get created in this directory
+        os.chdir(os.path.dirname(src_file))
+        my_buffer = StringIO()
+        my_stdout = Tee(sys.stdout, my_buffer)
+        sys.stdout = my_stdout
+
+        t_start = time()
+        exec(code_block, example_globals)
+        time_elapsed = time() - t_start
+
+        sys.stdout = orig_stdout
+
+        my_stdout = my_buffer.getvalue().strip().expandtabs()
+        if my_stdout:
+            stdout = CODE_OUTPUT.format(my_stdout.replace('\n', '\n    '))
+        os.chdir(cwd)
+        figure_list = save_figures(image_path, fig_count)
+
+        # Depending on whether we have one or more figures, we're using a
+        # horizontal list or a single rst call to 'image'.
+        if len(figure_list) == 1:
+            figure_name = figure_list[0]
+            image_list = SINGLE_IMAGE % figure_name.lstrip('/')
+        else:
+            image_list = HLIST_HEADER
+            for figure_name in figure_list:
+                image_list += HLIST_IMAGE_TEMPLATE % figure_name.lstrip('/')
+
+    except:
```

Comment:
Always use `except Exception`, or you capture `KeyboardInterrupt` making it hard for the user to break out of the whole process.